rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,222 @@
+
+With this effort, we want to provide an independent compaction/clustering Service, it will have these abilities:
+
+- Provides a pluggable execution interface that can adapt to multiple execution engines, such as Spark and Flink.
+
+- With the ability to failover, need to be persisted compaction/clustering message.
+
+- Perfect metrics and reuse HoodieMetric expose to the outside.
+
+- Provide automatic failure retry for compaction/clustering job.
+
+## Implementation
+
+![](service.png)
+
+### Client
+
+Register to Service with CompactionConfig/ClusteringConfig when the job starts. When generating a compaction/clustering plan, report to the compaction/clustering Service address specified in HoodieConfig and request the Service to stop the corresponding compaction/clustering job when the Service is not required or rollback compaction/clustering instant.
+
+### Request Handler
+
+Receive the client's request and save the plan in the meta table.
+
+### Meta Table
+
+Meta table has been implemented by using an internal HUDI MOR Table to store the required metadata for the compaction/clustering message. This table will be internal to a dataset and will not be exposed directly to the user to write/modify.
+
+The Meta Table has the following features:
+
+1. Is a HUDI MOR Table
+
+2. Defines its own custom payload HoodieServicePayload
+
+3. Use basePath + instant as recordKey to guarantee unique
+
+4. Follows all the conventions of a HUDI Table
+
+### Scheduler
+
+Periodically scan the Meta Table and submit compaction/clustering job according to user-specified rules, and need to plug-in the execution engine.
+
+### Schema
+
+Being an actual Hoodie Table, the Meta Table needs to have a schema to store the compaction/clustering message. 
+
+```
+{
+    "namespace":"org.apache.hudi.avro.model",
+    "type":"record",
+    "name":"HoodieServiceRecord",
+    "doc":"Compaction/Clustering message saved within the Meta Table",
+    "fields":[
+        {
+            "name":"table_name",
+            "type":"string"
+        },
+        {
+            "name":"base_path",
+            "type":"string"
+        },
+        {
+            "name":"instant",
+            "doc":"compaction/clustering instant for the hoodie table",
+            "type":"string"
+        }, {
+            "name":"action",
+            "doc":"compact or cluster for this instant",
+            "type":"string"
+        },{
+            "name":"status",
+            "doc":"Status of the compaction/clustering job, 0：Scheduled 1：Running 2：Retrying 3:  Completed",
+            "type":"int"
+        } ,{
+            "name":"isDeleted",
+            "doc":"True if this plan has been deleted",
+            "type":"int"
+        }
+    ]
+}
+```
+
+## Interface design
+
+### Register
+
+/v1/hoodie/service/register
+
+```
+{
+"table_name":"table_name",
+"base_path":"/hoodie/base_path",
+"compaction_config":"compaction_config",
+"clustering_config":"clustering_config"
+}
```

Comment:
Many other configs will also be required:
1. retry_on_error: true/false (if auto retry on error)
2. Periodic: true/false (if should be done periodically)
3. timeout: timeout for job execution
4. priority: some priority
5. context: some config which the user can add and retrieve. This helps us add "tier", "SLA", "owner", "yarn_queue" etc which are useful for the execution engine to schedule the job.

We have over 10000+ tables which belong to different teams. So scheduling jobs on each table should be executed in the table-owners yarn_queue.

Reply:
Don't we need entire write configs for each table as well.for eg, lock provider configurations?

Reply:
Yes, these parameters are necessary.